sos/plugins/juju.py

```diff
@@ -9,113 +9,201 @@
-            self.add_cmd_output('ls -alRh /var/log/juju*')
-            self.add_cmd_output('ls -alRh /var/lib/juju/*')
-
-        else:
-            self.add_copy_spec([
-                "/var/log/juju",
-                "/var/log/juju-*",
-                "/var/lib/juju"
-                # /var/lib/juju used to be in the default capture moving here
-                # because it usually was way to big.  However, in most cases
-                # you want all logs you want this too.
-            ])
-
-        self.add_cmd_output([
-                "juju --version",
-                "juju -v status --format=tabular",
-        ])
-        for service in self.get_deployed_services():
-            self.add_cmd_output([
-                "juju get {}".format(service),
-                "juju get-config {}".format(service),
-                "juju get-constraints {}".format(service)
-            ])
-
-        if self.get_option("export-mongodb"):
-            self.export_mongodb()
-
-        if self.get_option("generate-bundle"):
-            self.add_cmd_output("juju deployerizer --include-charm-versions",
-                                suggest_filename="juju-env-bundle.yaml")
+        # Make sure it looks like juju is configured before continuing
+        username = self.get_option("juju-user")
+        homedir = os.path.expanduser("~" + username)
+        if os.path.exists(homedir + "/.local/share/juju"):
+            self.collect_juju_output(username)
+
+    def postproc(self):
+        juju_config_keys = [
+            "password",
+            "credentials",
+            "client_password",
+            "endpoint-tls-ca",
+            "docker-logins",
+            "license-key",
+            "registry-credentials",
+        ]
+
+        juju_config_certs = ["ca-cert"]
+
+        # Handle the case where the key and value are on seperate lines
+        # in a juju config yaml file
+        juju_config_regex = (
+            r"(?m)^(\s*)(((%s):((\n\1\s+.*)|\n)+)\1\s+value:\s*).*"
+            % "|".join(juju_config_keys)
+        )
+        juju_config_sub_regex = r"\1\2*********"
+
+        # Will match and replace a certificat in a yaml file
+        certs_regex = (
+            r"((?m)^\s*(%s)\s*:\s*)(\|\s*\n\s+-+BEGIN (.*)"
+            r"-+\s(\s+\S+\n)+\s+-+END )\4(-+)" % "|".join(juju_config_certs)
+        )
+        certs_sub_regex = r"\1*********"
+
+        # Go through all output files and sanitize them (sos doesn't have a
+        # builtin method of doing this currently):
+        output_dir = self.get_cmd_output_path(make=False)
+        juju_config_files = glob.glob(output_dir + "/juju_config*.yaml")
+        juju_cert_files = glob.glob(output_dir + "/juju*controller*.yaml")
+
+        for file in juju_config_files:
+            self._log_debug(
+                "substituting {} for {} in {}".format(
+                    juju_config_sub_regex, juju_config_regex, file
+                )
+            )
```

Comment:
Why are you doing this? The postproc hooks *already* log what they are doing:

```python
 303     def do_file_sub(self, srcpath, regexp, subst):
 311         try:
 312             path = self._get_dest_for_srcpath(srcpath)
 313             self._log_debug("substituting scrpath '%s'" % srcpath)
 314             self._log_debug("substituting '%s' for '%s' in '%s'"
 315                             % (subst, regexp, path))
```

```python
 260     def do_cmd_output_sub(self, cmd, regexp, subst):
 272         self._log_debug("substituting '%s' for '%s' in commands matching '%s'"
 273                         % (subst, regexp, globstr))
 274 
```

These are infrastructure level messages. They don't belong in individual plugins.